sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java

```diff
@@ -78,6 +92,231 @@
+            };
+
+        final Predicate<Class<?>> disallowed = new Predicate<Class<?>>() {
+
+          @Override
+          public boolean apply(@Nonnull final Class<?> aClass) {
+            return !classIsAllowed(aClass, allowedClasses);
+          }
+        };
+
+    /* </helper_lambdas> */
+
+        final FluentIterable<Class<?>> disallowedClasses =
+            FluentIterable
+                .from(checkedApiSurface.getExposedClasses())
+                .filter(disallowed);
+
+        final ImmutableMap<Class<?>, List<Class<?>>> exposures =
+            Maps.toMap(disallowedClasses, toExposure);
+
+        final ImmutableList<String> messages =
+            FluentIterable
+                .from(Maps.transformEntries(exposures, toMessage).values())
+                .toSortedList(Ordering.<String>natural());
+
+        if (!messages.isEmpty()) {
+          mismatchDescription.appendText(
+              "The following disallowed classes appeared on the API surface:\n\t"
+                  + Joiner.on("\n\t").join(messages));
+        }
+
+        return messages.isEmpty();
+      }
+
+      @SuppressWarnings({ "rawtypes", "unchecked" })
+      private boolean classIsAllowed(final Class<?> clazz,
+                                     final Set<Matcher<Class<?>>> allowedClasses) {
+        // Safe cast inexpressible in Java without rawtypes
+        return anyOf((Iterable) allowedClasses).matches(clazz);
+      }
+
+      @Override
+      protected boolean matchesSafely(final ApiSurface apiSurface,
+                                      final Description mismatchDescription) {
+        final boolean noDisallowed =
+            verifyNoDisallowed(apiSurface, classMatchers, mismatchDescription);
+
+        final boolean noAbandoned =
+            verifyNoAbandoned(apiSurface, classMatchers, mismatchDescription);
+
+        return noDisallowed & noAbandoned;
+      }
+
+      @Override
+      public void describeTo(final Description description) {
+        description.appendText("API surface to include only:" + "\n\t");
+        for (final Matcher<Class<?>> classMatcher : classMatchers) {
+          classMatcher.describeTo(description);
+          description.appendText("\n\t");
+        }
+      }
+
+    }
+
+    /**
+     * A factory method to create an {@link ApiSurface} matcher, producing a positive match
+     * if the set of class matchers passed in is exposed by the queried api surface,
+     * AND
+     * none of the provided class matchers remains unmatched.
+     *
+     * @param classMatchers the class matchers that will produce a positive match for this matcher.
+     */
+    public static Matcher<ApiSurface> containsOnlyClassesMatching(
+        final Set<Matcher<Class<?>>> classMatchers) {
+      return new SurfaceMatcher(classMatchers);
+    }
+
+    /**
+     * See {@link ApiSurface.Matchers#containsOnlyClassesMatching(Set)}.
+     */
+    @SafeVarargs
+    public static Matcher<ApiSurface> containsOnlyClassesMatching(
```

Comment:
How about also `containsOnlyClassesInPackages(String...)` and `containsOnlyPackages(Iterable<? extends String>)`, which I think would be the most common uses anyhow.